### PR TITLE
Move one function from cuda_provider_factory.h

### PR DESCRIPTION
**Pull request by RyanUnderhill**
**Description**: Move OrtSessionOptionsAppendExecutionProvider_CUDA out of cuda_provider_factory.h and put it into the C API as it always exists now.

This removes the need to package this header file and for users to ever include it manually. It's now only used by our internal code.

### CODE REVIEW

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 1]
```diff
@@ -1502,6 +1502,14 @@ struct OrtCustomOp {
   OrtCustomOpInputOutputCharacteristic(ORT_API_CALL* GetOutputCharacteristic)(_In_ const struct OrtCustomOp* op, _In_ size_t index);
 };
 
+/*
+ * This is the old way to add the CUDA provider to the session, please use SessionOptionsAppendExecutionProvider_CUDA above to access the latest functionality
+ * This function always exists, but will only succeed if Onnxruntime was built with CUDA support and the CUDA provider shared library exists
+ * 
+ * \param device_id cuda device id, starts from zero.
+*/
+ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_CUDA, _In_ OrtSessionOptions* options, int device_id);
```
Comment (snnn) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:1511]: If it is deprecated, why do you add it here? If you want to keep backward compatibility, then you should make the old one continue to work.

Comment (RyanUnderhill) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:1511]: The old function still works fine, people just don't need to include the extra header to get it.

Comment (snnn) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:1511]: It is confusing.  We are adding this function to all EPs.
First, do we need to keep this? For what, backward compatibility? If it is important to the CUDA EP, then why it is not important to the other EPs? By doing this, you are breaking it for the others EPs.

Comment (RyanUnderhill) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:1511]: How does this break other EPs?

Comment (snnn) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:1511]: Ah, I was wrong. We are using def files, so this function won't be auto exported.

**Review by snnn — APPROVED** [review]

File: csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/C_Api_Sample.cpp [hunk 2]
```diff
@@ -39,7 +39,7 @@ int main(int argc, char* argv[]) {
   g_ort->SetSessionGraphOptimizationLevel(session_options, ORT_ENABLE_BASIC);
 
   // Optionally add more execution providers via session_options
-  // E.g. for CUDA include cuda_provider_factory.h and uncomment the following line:
+  // E.g. for CUDA uncomment the following line:
   // OrtSessionOptionsAppendExecutionProvider_CUDA(sessionOptions, 0);
```
Comment (pranavsharma) [inline csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/C_Api_Sample.cpp:43]: Shouldn't this be g_ort->OrtSessionOptionsAppendExecutionProvider_CUDA?

File: onnxruntime/python/tools/tensorrt/perf/mem_test/main.cpp [hunk 3]
```diff
@@ -160,9 +160,7 @@ void ort_trt_run_with_default_options() {
   Ort::SessionOptions session_options;
   session_options.SetIntraOpNumThreads(1);
 
-  // If onnxruntime.dll is built with CUDA enabled, we can uncomment out this line to use CUDA for this
-  // session (we also need to include cuda_provider_factory.h above which defines it)
-  // #include "cuda_provider_factory.h"
+  // If onnxruntime.dll is built with CUDA enabled, we can uncomment out this line to use CUDA for this session
   // OrtSessionOptionsAppendExecutionProvider_CUDA(session_options, 1);
```
Comment (pranavsharma) [inline onnxruntime/python/tools/tensorrt/perf/mem_test/main.cpp:164]: Needs update as commented before.

File: csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/C_Api_Sample.cpp (hunk 2 above)
Comment (RyanUnderhill) [reply csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/C_Api_Sample.cpp:43]: Yeah, it's probably a good idea to switch the samples to the latest API. But this was just updating the comment to be correct, The existing API still exists and works fine. It's just not as flexible as the new C API interface one.

Comment (snnn) [reply csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/C_Api_Sample.cpp:43]: I have a PR to delete the samples. https://github.com/microsoft/onnxruntime/pull/8441.  @pranavsharma, we discussed it.

**Pull request merged**